pyanaconda/modules/subscription/runtime.py

```diff
@@ -0,0 +1,115 @@
+    CONFIG_KEY_SERVER_PROXY_USER = "server.proxy_user"
+    CONFIG_KEY_SERVER_PROXY_PASSWORD = "server.proxy_password"
+    CONFIG_KEY_RHSM_BASEURL = "rhsm.baseurl"
+
+    def __init__(self, rhsm_config_proxy, rhsm_config_defaults, subscription_request):
+        """Create a new task for setting RHSM configuration.
+
+        :param rhsm_config_proxy: DBus proxy for the RHSM Config object
+        :param dict rhsm_config_defaults: a dictionary of original RHSM configuration values
+        :param subscription_request: subscription request DBus Structure
+        :type subscription_request: SubscriptionRequest instance
+        """
+        super().__init__()
+        self._rhsm_config_proxy = rhsm_config_proxy
+        self._request = subscription_request
+        self._rhsm_config_defaults = rhsm_config_defaults
+
+    @property
+    def name(self):
+        return "Set RHSM configuration."
+
+    def run(self):
+        log.debug("subscription: setting RHSM config values")
+        # We will use the SetAll() dbus method and we need to
+        # assemble a dictionary that we will feed to it.
+        # Start by preparing a SubscriptionData property mapping
+        # to the RHSM config keys.
+        #
+        # A note about constructing the dict:
+        # - DBus API needs all values to be strings, so we need to convert the
+        #   port number to string
+        # - proxy password is stored in SecretData instance and we need to retrieve
+        #   its value
+        property_key_map = {
+            self.CONFIG_KEY_SERVER_HOSTNAME: self._request.server_hostname,
+            self.CONFIG_KEY_SERVER_PROXY_HOSTNAME: self._request.server_proxy_hostname,
+            self.CONFIG_KEY_SERVER_PROXY_PORT: str(self._request.server_proxy_port),
+            self.CONFIG_KEY_SERVER_PROXY_USER: self._request.server_proxy_user,
+            self.CONFIG_KEY_SERVER_PROXY_PASSWORD: self._request.server_proxy_password.value,
+            self.CONFIG_KEY_RHSM_BASEURL: self._request.rhsm_baseurl
+        }
+
+        # Then process the mapping into the final dict we will set to RHSM. This includes
+        # checking if some values have been cleared by the user and should be restored to
+        # the original values that have been in the RHSM config before we started
+        # manipulating it.
+        config_dict = {}
+        for key, value in property_key_map.items():
+            if value:
+                # if value is present in request, use it
+                config_dict[key] = value
+            else:
+                # if no value is present in request, use
+                # value from the original RHSM config state
+                # (if any)
+                log.debug("subscription: restoring original value for RHSM config key %s", key)
+                config_dict[key] = self._rhsm_config_defaults.get(key, "")
+
+        # turn the dict into variant
+        config_variant = get_variant(Dict[Str, Str], config_dict)
```

Comment:
It needs to be a dictionary that maps strings to variants, so you need to create a string variant for every value in the dictionary.

Reply:
This seems like something that we could have a generic function for, that would take the values in a Python dict and turn them to appropriate variants - "foo" -> String variant, 1 -> Int variant, etc.

It would be basically the inverse of `get_native()`.

Reply:
Anyway, switched the `SetAll()` call to use a regular dict that has string keys and string variant values - hopefully got it right this time. :P

Reply:
> This seems like something that we could have a generic function for, that would take the values in a Python dict and turn them to appropriate variants - "foo" -> String variant, 1 -> Int variant, etc.

We are not able to do that, because 1 can be also Byte, Int16, UInt16, Int32, UInt32, Int64, UInt64. This is the reason why we have DBus structures.